### ISSUE TITLE
Why is access control code inbox-centric?

### ISSUE THREAD

**Issue by fsalomon** [chatwoot/chatwoot#4964]
Hi,

while scrolling through the code I've had trouble understanding why you chose to make just about all access control code inbox-centric, e.g. delegating conversation access directly to InboxPolicy, bypassing ConversationPolicy altogether. I understand that you currently have access _config_ based around inboxes but that does not mean you should make your _code_ follow this assumption. Even worse when such assumptions creep into frontend-code e.g. ``/api/v1/accounts/.../assignable_agents?inbox_ids[]=...`` should imho be ``/api/v1/accounts/.../conversations/.../assignable_agents``. Such abstraction would future-proof your code, because I find it inevitable that you will eventually move away from this inbox-centric config.

This brings me to my next point: Why did you choose to make access _config_ inbox-centric? Assuming we have an account where access is not "free for all"; I find it hard to come up with a scenario where inbox-centric config adequately reflects the real-life responsibilities of agents. Maybe I'm missing something essential here, but customers are notorious to pick their contact channel on a least-effort basis and not putting much thought into how to reach the right department/team.

In a traditional phone support system it is certainly possible to route a call to the right department/team even if the call did not initially reach a qualified agent. Customers expect the same when they're using a text-based support channel. Now let's say that for whatever reason we do not want the billing team to have access to _all_ of the company's Twitter DMs. However, there will certainly be customers asking billing related questions over Twitter DM's. So there must be a way to transfer a conversation over to an agent that does not have full access to the Twitter inbox. Conversation->team assignment is pointing in the right direction but that does not currently transcend inbox boundaries and the hard-coded assumptions mentioned above make it hard to implement a more flexible access control.

Can somebody enlighten me what real-world scenario drove you to make inboxes the first-class citizen of access control within an account?

**Comment by sojan-official** [comment]
@fsalomon, the reasons are primarily historical. If you look at the story of chatwoot ( ref: https://www.chatwoot.com/docs/handbook/about-chatwoot/history#2016-the-beginning-of-chatwoot ), The open-source code base evolved from the original version, which was meant to manage multiple Facebook pages under a single dashboard, where we had inbox centric permissions. Even as the product involved, the foundation remained same as it is now harder to change the workflows with out it breaking for current users. 

But we hear you point and can see how it will evolve in the future to be more account centric. Lets keep this issue open to see further discussions from the community. 

also linking some of the issues which we are tracking to improve the behaviour around permissions: 

https://github.com/chatwoot/chatwoot/issues/2710
https://github.com/chatwoot/chatwoot/issues/4216
https://github.com/chatwoot/chatwoot/issues/1588
https://github.com/chatwoot/chatwoot/issues/2443

**Comment by fsalomon** [comment]
Yes, I figured this might be legacy related. I suspect that a large percentage of accounts are actually configured "free for all" i.e. all user access have access to all inboxes; indicating a clear misalignment between access config and real-world usage.

My issue is mostly about refactoring, the changes can come in later. For example I found 
```
  # Deprecated: This API will be removed in 2.7.0
  def assignable_agents
    @assignable_agents = (Current.account.users.where(id: @inbox.members.select(:user_id)) + Current.account.administrators).uniq
  end
```

in `InboxesController`. So it seems you are already doing some API changes. But moving the API to `AssignableAgentsController` is not the right way to go imho. This functionality should be in `ConversationsController` so that the frontend can be agnostic as to how access control is handled in the backend.

**Comment by fsalomon** [comment]
It all comes down to the migration path between where access control is now and where it might be in the future. Keeping your frontend as agnostic as possible should make that easier. 

For what it's worth, I will post how I "fixed" access control for my use case. I'm eager to hear what others think about this.
- So my scenario requires certain agents (actually a team of agents but that is just a small admin nuisance) to have direct access to some of the inboxes. Let's call this "full agent"
- There are also other agents which should only have access to conversations assigned to their team. Let's call this "restricted agent"

My goal was to do this without needing code changes to the mobile app, so I had to remove most of the inbox-level access control code from the backend. So every agent will see all inboxes but the restricted agents can only see their team's conversation in any of the inboxes. At the core of my fix is the following code in `ConversationFinder`. I added some comments for further clarification. Keep in mind that my total lifetime exposure to Rails is less than 6 hours, so maybe I made some terrible mistake :-)
```
  def find_all_conversations
    if params[:conversation_type] == 'mention'
      conversation_ids = current_account.mentions.where(user: current_user).pluck(:conversation_id)
      @conversations = current_account.conversations.where(id: conversation_ids)
    elsif params[:inbox_id]
      # user is trying to access a specific inbox
      begin
        # check if the user has direct access to the inbox
        @inbox_ids = @current_user.assigned_inboxes.find(params[:inbox_id])
        @conversations = current_account.conversations.where(inbox_id: @inbox_ids)
      rescue ActiveRecord::RecordNotFound => e
        # the user does not have direct access to the inbox, so we will show them only conversations assigned to one of their teams
        @inbox_ids = @current_account.inboxes.find(params[:inbox_id])
        @team_ids = @current_user.assigned_teams.pluck(:id)
        @conversations = current_account.conversations.where(inbox_id: @inbox_ids, team_id: @team_ids)
      end
    elsif params[:team_id]
      # user is trying to access a specific team
      @team_ids = @current_user.assigned_teams.find(params[:team_id])
      @conversations = current_account.conversations.where(team_id: @team_ids)
    else
      # user is trying to access "all conversations"
      # showing all conversations that are either in an inbox the user has direct access to or are assigned to one of their teams
      @inbox_ids = @current_user.assigned_inboxes.pluck(:id)
      @team_ids = @current_user.assigned_teams.pluck(:id)
      @conversations = current_account.conversations.where(inbox_id: @inbox_ids).or(current_account.conversations.where(team_id: @team_ids))
    end
  end
```

The beauty of this is that it requires zero code changes in the frontend or mobile app and zero changes to the database schema, no additional "roles" etc.

**Comment by fsalomon** [comment]
https://github.com/chatwoot/chatwoot/issues/2710: This is a workflow nuisance I have as well but I consider it low priority because it can be fixed with a script or an human chore
https://github.com/chatwoot/chatwoot/issues/4216: Would require a mayor overhaul of the access control, should be done in steps given how rudimentary access control is now
https://github.com/chatwoot/chatwoot/issues/1588: Seems purely cosmetic/workflow related
https://github.com/chatwoot/chatwoot/issues/2443: The more fine-grained access control and tagging the better, obliviously. But as I see it, my proposed refactoring would be a prerequisite for any of those changes. Inbox-centric access control has to be retired. I propose to keep the agent<->inbox assignment UI; but re-purpose it to mean "direct inbox access" as described above
